### ISSUE TITLE
PDF with huge embedded image is displayed as a blank page

### ISSUE THREAD

**Issue by Rob--W** [mozilla/pdf.js#6741]
PDF: [DC000933-uncompressed.pdf](https://github.com/mozilla/pdf.js/files/57455/DC000933-uncompressed.pdf)

If I open the above PDF in Chrome or Firefox, it is shown as blank at best. Firefox sometimes crashes if I (re)load a couple of times (even at a fresh start-up). I haven't built Firefox with symbols, but GDB shows that the stack trace ends in `mozalloc_abort`, which suggets that this segfault is caused by a memory allocation error.

This PDF is quite unusual, it has a width and height with excessive values:

```
<< /Contents 10 0 R /MediaBox [ 0 0 4444 2592 ] /Parent 6 0 R /Resources << /ExtGState << /G0 11 0 R /G1 12 0 R /G2 13 0 R >> /ProcSet [ /ImageB /ImageC /ImageI ] /XObject << /I0 14 0 R >> >> /Type /Page >>

...

<< /BitsPerComponent 1 /ColorSpace /DeviceGray /DecodeParms << /Columns 24688 /K -1 /Rows 14400 >> /Filter /CCITTFaxDecode /Height 14400 /Name /I0 /Subtype /Image /Type /XObject /Width 24688 /Length 972551 >>
```

It was allegedly created by MetaPrint.

**Comment by timvandermeij** [comment]
Yes, we have seen such PDF files before with images with a huge width and height. We might need to implement something to recognize that and downscale the image.

**Comment by AnasAl-tamemy** [comment]
i think it should return an error so that developers can handle the open pdf failure.

**Comment by gitarra** [comment]
These files render fine at least with Evince and PDFium.

**Comment by redlum94** [comment]
Is there already any news on this issue? We run into the same issue with half of our PDFs. As they consist of scanned building plans. Or is there a known cause of this problem?

**Comment by Rob--W** [comment]
@redlum94 Scanned building plans sound like large images (especially with a high resolution). In the first comment, @timvandermeij already suggested that we should downscale the image when needed.

I have worked on image downscaling before in PR #6606 , but that patch has bitrotten.

**Comment by dylanwoo** [comment]
Does anyone know if there is a global variable I can look at to see if the viewer is empty or not?

**Comment by peterkmurphy** [comment]
I am another user affected by the bug. My tool also works with architectural diagrams, where pages sizes can be A1 or greater in size.

This bug was triggered by a PDF which was 841 (width) × 1007 (height) mm in size. What is interesting is that pdf.js handles A0 Landscape PDFs (1189 × 841 mm) and A0 Portrait PDFs (841 × 1189 mm) no problem. So it is not just a case of PDFs being "too big". Certain dimensions cause problems - even those smaller than other dimensions that don't cause problems.

**Comment by THausherr** [comment]
The issue here related to the size of images within the PDF, not the size of the PDF. A good architectural PDF would use vector graphics, so the problem here wouldn't occur at all for such files.

**Comment by peterkmurphy** [comment]
Is that the case, @THausherr? I experimented with two A0 Landscape pdfs containing one image each (with no vector graphics whatsoever), and pdf.js handled it fine on my computer. It just choked on the 841 x 1007 mm pdf.

**Comment by THausherr** [comment]
It also depends on the size of the image within the PDF. The image from the PDF mentioned on top is 14400 x 24688. It would be interesting to see the size and colorspace of your PDFs.

**Comment by peterkmurphy** [comment]
I've got a lot to learn, obviously. Opening the 841 x 1007 PDF in a text editor, I get "/Width 13248 /Height 15858 /ColorSpace /DeviceGray". The A0 PDF has "/ColorSpace/DeviceRGB /Width 6622 /Height 9362". I hope that answers your question, @THausherr .

**Comment by jxintang** [comment]
so is there already any news on this issue?

**Comment by mlongo4290** [comment]
I'm experiencing the same bug with files scanned at 1200dpi, so i can confirm what @THausherr said, the scanner makes a one page pdf with a image sized 9824x13934. The color space is "/DeviceGray" but I think that the main issue here is the picture size.

**Comment by jxintang** [comment]
change option.maxImageSize can solve this problem.

pdfjsLib.getDocument({
      url: url,
      maxImageSize: MAX_IMAGE_SIZE

**Comment by mlongo4290** [comment]
@jxintang it does not work for me. Besides, documentation reports maxImageSize with a default value of -1 which means "renderer everything" but does not solve any issue.
` * @property {number} maxImageSize - (optional) The maximum allowed image size
 *   in total pixels, i.e. width * height. Images above this value will not be
 *   rendered. Use -1 for no limit, which is also the default value.`
Am I wrong?

**Comment by jxintang** [comment]
> @jxintang it does not work for me. Besides, documentation reports maxImageSize with a default value of -1 which means "renderer everything" but does not solve any issue.
> ` * @Property {number} maxImageSize - (optional) The maximum allowed image size
> 
> * in total pixels, i.e. width * height. Images above this value will not be
> * rendered. Use -1 for no limit, which is also the default value.`
>   Am I wrong?

in my case, i can see a warning in my chrome like this
`Warning: Image exceeded maximum allowed size and was removed.`
so i just check the source code in pdf.worker.js
![image](https://user-images.githubusercontent.com/5434292/64531174-6ad21a80-d341-11e9-865f-72074c2fe0a1.png)

when i change the option.maxImageSize, the problem solved

**Comment by pklion** [comment]
I have same issue on my FF, but will it be fixed? I think this makes pdf.js not usable...

**Comment by Benguenette** [comment]
Hi,
I come and read here every once in a while hoping that a solution shows up. It's been a few years now and working in engineering document management like some of the users here, It gives us a headache.

Is there an explanation why it's still open or maybe there's a limitation in the code making a solution impossible? It would be really appreciated.

**Comment by 99991** [comment]
[My issue on bugzilla from two years ago](https://bugzilla.mozilla.org/show_bug.cgi?id=1646282) got closed as a duplicate of [this issue from 6 months ago,](https://bugzilla.mozilla.org/show_bug.cgi?id=1713266) which references this GitHub issue, so I would like to link my PDF document here as well, hoping that it might help to resolve this issue.

https://bugzilla.mozilla.org/attachment.cgi?id=9157210

> The issue here is related to the size of images within the PDF, not the size of the PDF. A good architectural PDF would use vector graphics, so the problem here wouldn't occur at all for such files.

This PDF does not contain any raster graphics. Instead, it contains a (large) SVG file, yet it does not render completely unless zoomed out (the bottom part is missing).

**Comment by THausherr** [comment]
That is a different problem than this one IMHO. Your PDF has a huge Y size. I see it partially in Firefox on window, until 171, and it's blurry.

**Comment by Snuffleupagus** [comment]
> https://bugzilla.mozilla.org/attachment.cgi?id=9157210

> That is a different problem than this one IMHO. Your PDF has a huge Y size. I see it partially in Firefox on window, until 171, and it's blurry.

In that case, this particular document would rather be a "duplicate" of issue #6419 instead.

**Comment by MaRieck** [comment]
I appreciate that some movement comes into this :) hopefully you are able to find it @marco-c and @brendandahl .BTW: I think this "performance" label is wrong. This issue has nothing to do with performance but rather of functionality/bug in edge-cases (embedded image size too large).

**Comment by mraonea** [comment]
Practically all planning documents distributed as a PDF in the world are scanned and therefore images embedded into the PDF file and Firefox cannot display these. Add insult to injury, even after years of having this issue known, Firefox still just silently fails with these files without any hint to the user like "open in a proper PDF viewer like Chrome".

The funny thing is that viewing actual vector documents is not even a solution to this issue due to the abysmal performance of PDF.js. So, again, "open with Chrome".

**Comment by basantmatharu** [comment]
Is there any update on this issue? Our customers have some PDFs with huge embedded images and they fail to render without any warnings. I have not modified the maxImageSize property.
I did not find any alternative to PDF.js also for rendering PDFs programmatically.

If anyone has found an alternative or a workaround, please suggest

**Comment by Benguenette** [comment]
I second @basantmatharu.
This is an issue we are facing and would really love to see a work around.

**Comment by rschoch** [comment]
@basantmatharu what i did was try rendering with progressively smaller viewports. something like this:

```
    const renderTask = page.render({
      canvasContext: context,
      viewport,
    });
    renderTask.promise
      .then(() => {
        const jpegDataURL = canvas.toDataURL('image/jpeg', 1);
        resolve(jpegDataURL);
      })
      .catch(() => {
        if (scaleFactor && scaleFactor > 1) {
          resolve(renderPDFToDataUrl(typedArray, scaleFactor - 0.5));
        } else {
          reject(new Error(RenderError.RENDER_FAILED));
        }
      });
```

having said that, I start with scale factor 3 or 4 and "work my way down" until it renders successfully.

**Comment by basantmatharu** [comment]
@rschoch can you please share a gist with more code, if that's possible? Thanks

**Comment by devalberg** [comment]
Is there any way of detecting/catching this error (or any similar errors)? I've tried searching everywhere but no luck...

The idea is to render an image of the pdf rendered by an API using PDFium as a fallback, but without a way of detecting this it feels impossible.

The only "error" I'm getting is a warning from canvas.  
I've tried `stopAtErrors` + `maxImageSize` options in `.getDocument()` - but that doesn't throw anything, just stops rendering.

Am I missing something?
